File: data/src/main/java/org/cryptomator/data/cloud/webdav/network/WebDavCompatibleHttpClient.java
package org.cryptomator.data.cloud.webdav.network;

import static com.google.common.net.HttpHeaders.CACHE_CONTROL;
import static org.cryptomator.data.util.NetworkTimeout.CONNECTION;
import static org.cryptomator.data.util.NetworkTimeout.READ;
import static org.cryptomator.data.util.NetworkTimeout.WRITE;
import static org.cryptomator.util.file.LruFileCacheUtil.Cache.WEBDAV;

import java.io.IOException;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;

import javax.net.ssl.X509TrustManager;

import org.cryptomator.data.cloud.okhttplogging.HttpLoggingInterceptor;
import org.cryptomator.domain.WebDavCloud;
import org.cryptomator.domain.exception.UnableToDecryptWebdavPasswordException;
import org.cryptomator.util.SharedPreferencesHandler;
import org.cryptomator.util.crypto.CredentialCryptor;
import org.cryptomator.util.file.LruFileCacheUtil;

import com.burgstaller.okhttp.AuthenticationCacheInterceptor;
import com.burgstaller.okhttp.CachingAuthenticatorDecorator;
import com.burgstaller.okhttp.DispatchingAuthenticator;
import com.burgstaller.okhttp.basic.BasicAuthenticator;
import com.burgstaller.okhttp.digest.CachingAuthenticator;
import com.burgstaller.okhttp.digest.Credentials;
import com.burgstaller.okhttp.digest.DigestAuthenticator;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import okhttp3.Authenticator;
import okhttp3.Cache;
import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import timber.log.Timber;

class WebDavCompatibleHttpClient {

	private final WebDavRedirectHandler webDavRedirectHandler;

	WebDavCompatibleHttpClient(WebDavCloud cloud, Context context) {
		final SharedPreferencesHandler sharedPreferencesHandler = new SharedPreferencesHandler(context);
		this.webDavRedirectHandler = new WebDavRedirectHandler(httpClientFor(cloud, context, sharedPreferencesHandler.useLruCache(), sharedPreferencesHandler.lruCacheSize()));
	}

	Response execute(Request.Builder requestBuilder) throws IOException {
		return execute(requestBuilder.build());
	}

	private Response execute(Request request) throws IOException {
		return webDavRedirectHandler.executeFollowingRedirects(request);
	}

	private static OkHttpClient httpClientFor(WebDavCloud webDavCloud, Context context, boolean useLruCache, int lruCacheSize) {
		final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();

		OkHttpClient.Builder builder = new OkHttpClient() //
				.newBuilder() //
				.connectTimeout(CONNECTION.getTimeout(), CONNECTION.getUnit()) //
				.readTimeout(READ.getTimeout(), READ.getUnit()) //
				.writeTimeout(WRITE.getTimeout(), WRITE.getUnit()) //
				.followRedirects(false) //
				.addInterceptor(httpLoggingInterceptor(context)) //
				.authenticator(httpAuthenticator(context, webDavCloud, authCache)) //
				.addInterceptor(new AuthenticationCacheInterceptor(authCache));

		if (useLruCache) {
			final Cache cache = new Cache(new LruFileCacheUtil(context).resolve(WEBDAV), lruCacheSize);
			builder.cache(cache) //
					.addNetworkInterceptor(provideCacheInterceptor()) //
					.addInterceptor(provideOfflineCacheInterceptor(context));
		}

		X509TrustManager trustManager;
		if (usingWebDavWithSelfSignedCertificate(webDavCloud)) {
			PinningTrustManager pinningTrustManager = new PinningTrustManager(webDavCloud.certificate());
			trustManager = pinningTrustManager;
			builder.hostnameVerifier(pinningTrustManager.hostnameVerifier());
		} else {
			trustManager = new DefaultTrustManager();
		}
		builder.sslSocketFactory(SSLSocketFactories.from(trustManager), trustManager);

		return builder.build();
	}

	private static Interceptor provideOfflineCacheInterceptor(final Context context) {
		return chain -> {
			Request request = chain.request();

			if (isNetworkAvailable(context)) {
				final CacheControl cacheControl = new CacheControl.Builder() //
						.maxAge(0, TimeUnit.DAYS) //
						.build();

				request = request.newBuilder() //
						.cacheControl(cacheControl) //
						.build();
			}

			return chain.proceed(request);
		};
	}

	private static Interceptor provideCacheInterceptor() {
		return chain -> {
			final Response response = chain.proceed(chain.request());
			final CacheControl cacheControl = new CacheControl.Builder() //
					.maxAge(0, TimeUnit.DAYS) //
					.build();

			return response.newBuilder() //
					.removeHeader("Pragma") //
					.removeHeader("Cache-Control") //
					.header(CACHE_CONTROL, cacheControl.toString()) //
					.build();
		};
	}

	private static Authenticator httpAuthenticator(Context context, WebDavCloud webDavCloud, Map<String, CachingAuthenticator> authCache) {
		Credentials credentials = new Credentials(webDavCloud.username(), decryptPassword(context, webDavCloud.password()));
		final BasicAuthenticator basicAuthenticator = new BasicAuthenticator(credentials);
		final DigestAuthenticator digestAuthenticator = new DigestAuthenticator(credentials);

		Authenticator result = new DispatchingAuthenticator //
				.Builder() //
						.with("digest", digestAuthenticator) //
						.with("basic", basicAuthenticator) //
						.build();
		result = new CachingAuthenticatorDecorator(result, authCache);

		return result;
	}

	private static String decryptPassword(Context context, String password) throws UnableToDecryptWebdavPasswordException {
		try {
			return CredentialCryptor //
					.getInstance(context) //
					.decrypt(password);
		} catch (RuntimeException e) {
			throw new UnableToDecryptWebdavPasswordException(e);
		}
	}

	private static Interceptor httpLoggingInterceptor(Context context) {
		return new HttpLoggingInterceptor(message -> Timber.tag("OkHttp").d(message), context);
	}

	private static boolean usingWebDavWithSelfSignedCertificate(WebDavCloud webDavCloud) {
		return webDavCloud.certificate() != null;
	}

	private static boolean isNetworkAvailable(final Context context) {
		ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
		NetworkInfo activeNetworkInfo = connectivityManager.getActiveNetworkInfo();
		return activeNetworkInfo != null && activeNetworkInfo.isConnected();
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/webdav/network/WebDavCompatibleHttpClient.kt
package org.cryptomator.data.cloud.webdav.network

import android.content.Context
import android.net.ConnectivityManager
import com.burgstaller.okhttp.AuthenticationCacheInterceptor
import com.burgstaller.okhttp.CachingAuthenticatorDecorator
import com.burgstaller.okhttp.DispatchingAuthenticator
import com.burgstaller.okhttp.basic.BasicAuthenticator
import com.burgstaller.okhttp.digest.CachingAuthenticator
import com.burgstaller.okhttp.digest.Credentials
import com.burgstaller.okhttp.digest.DigestAuthenticator
import com.google.common.net.HttpHeaders
import org.cryptomator.data.cloud.okhttplogging.HttpLoggingInterceptor
import org.cryptomator.data.util.NetworkTimeout
import org.cryptomator.domain.WebDavCloud
import org.cryptomator.domain.exception.UnableToDecryptWebdavPasswordException
import org.cryptomator.util.SharedPreferencesHandler
import org.cryptomator.util.crypto.CredentialCryptor
import org.cryptomator.util.crypto.FatalCryptoException
import org.cryptomator.util.file.LruFileCacheUtil
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import okhttp3.Authenticator
import okhttp3.Cache
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import timber.log.Timber

internal class WebDavCompatibleHttpClient(cloud: WebDavCloud, context: Context) {

	private val webDavRedirectHandler: WebDavRedirectHandler

	@Throws(IOException::class)
	fun execute(requestBuilder: Request.Builder): Response {
		return execute(requestBuilder.build())
	}

	@Throws(IOException::class)
	private fun execute(request: Request): Response {
		return webDavRedirectHandler.executeFollowingRedirects(request)
	}

	companion object {

		private fun httpClientFor(webDavCloud: WebDavCloud, context: Context, useLruCache: Boolean, lruCacheSize: Int): OkHttpClient {
			val authCache: Map<String, CachingAuthenticator> = ConcurrentHashMap()

			val builder = OkHttpClient() //
				.newBuilder() //
				.connectTimeout(NetworkTimeout.CONNECTION.timeout, NetworkTimeout.CONNECTION.unit) //
				.readTimeout(NetworkTimeout.READ.timeout, NetworkTimeout.READ.unit) //
				.writeTimeout(NetworkTimeout.WRITE.timeout, NetworkTimeout.WRITE.unit) //
				.followRedirects(false) //
				.addInterceptor(httpLoggingInterceptor(context)) //
				.authenticator(httpAuthenticator(context, webDavCloud, authCache)) //
				.addInterceptor(AuthenticationCacheInterceptor(authCache))
				.addInterceptor(UserAgentInterceptor())

			if (useLruCache) {
				val cache = Cache(LruFileCacheUtil(context).resolve(LruFileCacheUtil.Cache.WEBDAV), lruCacheSize.toLong())
				builder.cache(cache) //
					.addNetworkInterceptor(provideCacheInterceptor()) //
					.addInterceptor(provideOfflineCacheInterceptor(context))
			}

			val trustManager = if (usingWebDavWithSelfSignedCertificate(webDavCloud)) {
				val pinningTrustManager = PinningTrustManager(webDavCloud.certificate())
				builder.hostnameVerifier(pinningTrustManager.hostnameVerifier())
				pinningTrustManager
			} else {
				DefaultTrustManager()
			}
			builder.sslSocketFactory(SSLSocketFactories.from(trustManager), trustManager)

			return builder.build()
		}

		private fun provideOfflineCacheInterceptor(context: Context): Interceptor {
			return Interceptor { chain: Interceptor.Chain ->
				var request = chain.request()
				if (isNetworkAvailable(context)) {
					val cacheControl = CacheControl.Builder() //
						.maxAge(0, TimeUnit.DAYS) //
						.build()
					request = request.newBuilder() //
						.cacheControl(cacheControl) //
						.build()
				}
				chain.proceed(request)
			}
		}

		private fun provideCacheInterceptor(): Interceptor {
			return Interceptor { chain: Interceptor.Chain ->
				val response = chain.proceed(chain.request())
				val cacheControl = CacheControl.Builder() //
					.maxAge(0, TimeUnit.DAYS) //
					.build()
				response.newBuilder() //
					.removeHeader("Pragma") //
					.removeHeader("Cache-Control") //
					.header(HttpHeaders.CACHE_CONTROL, cacheControl.toString()) //
					.build()
			}
		}

		private fun httpAuthenticator(context: Context, webDavCloud: WebDavCloud, authCache: Map<String, CachingAuthenticator>): Authenticator {
			val credentials = Credentials(webDavCloud.username(), decryptPassword(context, webDavCloud.password()))
			val basicAuthenticator = BasicAuthenticator(credentials, StandardCharsets.UTF_8)
			val digestAuthenticator = DigestAuthenticator(credentials)
			var result: Authenticator = DispatchingAuthenticator.Builder() //
				.with("digest", digestAuthenticator) //
				.with("basic", basicAuthenticator) //
				.build()
			result = CachingAuthenticatorDecorator(result, authCache)
			return result
		}

		@Throws(UnableToDecryptWebdavPasswordException::class)
		private fun decryptPassword(context: Context, password: String): String {
			return try {
				CredentialCryptor //
					.getInstance(context) //
					.decrypt(password)
			} catch (e: FatalCryptoException) {
				throw UnableToDecryptWebdavPasswordException(e)
			}
		}

		private fun httpLoggingInterceptor(context: Context): Interceptor {
			val logger = object : HttpLoggingInterceptor.Logger {
				override fun log(message: String) {
					Timber.tag("OkHttp").d(message)
				}
			}

			return HttpLoggingInterceptor(logger, context)
		}

		private fun usingWebDavWithSelfSignedCertificate(webDavCloud: WebDavCloud): Boolean {
			return webDavCloud.certificate() != null
		}

		private fun isNetworkAvailable(context: Context): Boolean {
			val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
			val activeNetworkInfo = connectivityManager.activeNetworkInfo
			return activeNetworkInfo != null && activeNetworkInfo.isConnected
		}
	}

	class UserAgentInterceptor : Interceptor {

		@Throws(IOException::class)
		override fun intercept(chain: Interceptor.Chain): Response {
			val originalRequest: Request = chain.request()
			val userAgent = "davfs2/1.5.2"
			val requestWithUserAgent = originalRequest.newBuilder().header("User-Agent", userAgent).build()
			return chain.proceed(requestWithUserAgent)
		}
	}

	init {
		val sharedPreferencesHandler = SharedPreferencesHandler(context)
		webDavRedirectHandler = WebDavRedirectHandler(httpClientFor(cloud, context, sharedPreferencesHandler.useLruCache(), sharedPreferencesHandler.lruCacheSize()))
	}
}
